src/ui/EditorNS/editor.cpp

```diff
@@ -247,56 +242,60 @@ namespace EditorNS
                 .get().toInt();
     }
 
-    QList<QMap<QString, QString>> Editor::languages()
+    void Editor::setLanguage(const Language* lang)
     {
-        QMap<QString, QVariant> languages =
-                asyncSendMessageWithResult("C_FUN_GET_LANGUAGES").get().toMap();
-
-        QList<QMap<QString, QString>> out;
-
-        QMap<QString, QVariant>::iterator lang;
-        for (lang = languages.begin(); lang != languages.end(); ++lang) {
-            QMap<QString, QVariant> mode = lang.value().toMap();
-
-            QMap<QString, QString> newMode;
-            newMode.insert("id", lang.key());
-            newMode.insert("name", mode.value("name").toString());
-            newMode.insert("mime", mode.value("mime").toString());
-            newMode.insert("mode", mode.value("mode").toString());
-
-            out.append(newMode);
+        if (!m_customIndentationMode) {
+            setIndentationMode(lang->id);
         }
-
-        return out;
+        m_currentLanguage = lang;
```

Comment:
Return early if `(m_currentLanguage == lang)` so we don't emit signals/etc.

Reply:
And there's still the chance of `lang == nullptr`. In that case we should probably revert back to plaintext (or defaultLanauge).